quil-rs/src/program/mod.rs

```diff
@@ -261,11 +276,76 @@ impl Program {
             used_qubits: HashSet::new(),
         };
 
-        new_program.add_instructions(expanded_instructions);
+        for (index, instruction) in self.instructions.iter().enumerate() {
+            match self.calibrations.expand_with_detail(instruction, &[])? {
+                Some(expanded) => {
+                    new_program._append_calibration_expansion_output(
+                        expanded,
+                        index,
+                        &mut source_mapping,
+                    );
+                }
+                None => {
+                    new_program.add_instruction(instruction.clone());
+                    if let Some(source_mapping) = source_mapping.as_mut() {
+                        source_mapping.entries.push(SourceMapEntry {
+                            source_location: index,
+                            target_location: MaybeCalibrationExpansion::Unexpanded(
+                                new_program.instructions.len() - 1,
+                            ),
+                        });
+                    }
+                }
+            }
+        }
 
         Ok(new_program)
     }
 
+    /// Append the result of a calibration expansion to this program, being aware of which expanded instructions
+    /// land in the program body (and thus merit inclusion within a target range) and which do not.
+    ///
+    /// For example, `DECLARE` instructions are hoisted to a specialized data structure and thus do not appear in
+    /// the program body. Thus, they should not be counted in the `target_index` range within a [`SourceMapEntry`].
+    fn _append_calibration_expansion_output(
+        &mut self,
+        mut expansion_output: CalibrationExpansionOutput,
+        source_index: usize,
+        source_mapping: &mut Option<&mut ProgramCalibrationExpansionSourceMap>,
+    ) {
+        if let Some(source_mapping) = source_mapping.as_mut() {
+            let previous_program_instruction_body_length = self.instructions.len();
+
+            for instruction in expansion_output.new_instructions {
+                let start_length = self.instructions.len();
+                self.add_instruction(instruction.clone());
+                let end_length = self.instructions.len();
+
+                // If the instruction was not added to the program body, remove its target index from the source map
+                // so that the map stays correct.
+                if start_length == end_length {
+                    let relative_target_index =
+                        start_length - previous_program_instruction_body_length;
+                    expansion_output
+                        .detail
+                        .remove_target_index(relative_target_index);
```

Comment:
This feels very clumsy and likely inefficient, but no more clever way occurred to me. Suggestions welcome.

Basically, we determine whether an instruction is a "header" or "body" instruction by whether or not it made the instruction body longer, and then based on that we excise it from the source mapping so that the `target_index` ranges stay accurate and don't map to a missing instruction

Reply:
I don't necessarily have a better idea for this, but an observation that the distinction between "header" and "body" instruction comes up enough that maybe we should have a more concrete definition of it in the library? Like a method in `Instruction` maybe? I know it's not a part of the spec, but it is a practical one that this library makes decisions around.

Reply:
> if the calibration expansion can check if something is a "header" instruction, then it can refrain from adding it to the expanded source, and then we don't need to do this fixup.

I don't see how this is possible. When the instruction is expanded, it does not have the scope of a program, and so it is just expanded into a `Vec<Instruction>`. However, when we add that Vec to a program here, then some of its members are "hoisted" into data structures like memory region declarations and calibration declarations and such. Expanding the calibration in such a way that "predicted" how the program would handle it, without the use of an actual program, is something that could fall out of sync and requires extra care and maintenance.

Reply:
It's statically knowable which instructions get hoisted and which ones don't: `Instruction::CalibrationDefinition` does, `Instruction::Gate` doesn't, etc.  Expansion could return *two* vectors: `Vec<HoistedInstruction>` and `Vec<ComputationInstruction>` (new types optional, but perhaps clearer).  Then anything working on the results of expansion wouldn't have to do its own classification.